### PR TITLE
[LoginNodes] Allow multiple login node pools

### PR DESCRIPTION
**Pull request by cjmakin**
### Description of changes
This change allows users to deploy clusters with multiple login node pools.

* Removed validator restricting number of pools to 1.
* Added duplicate name validator to `LoginNodes` resource to check that pool names are unique.
* Refactored `login_nodes_status` to support multiple pools.
    * Previously only configured to get information of the first pool. Now retrieves information for all pools in the cluster. This is used in the `update_policy.py` and `cluster_operations_controller.py` to get both total login nodes and pool specific statuses.
* Added `LOGIN_NODES_POOL_STOP` update policy.
    * The policy requires that all nodes in the pool in which the update is taking place are terminated. The previous policy, `LOGIN_NODES_STOP`, still remains and checks that all login nodes in the cluster are terminated. 
    * Changed each pool specific field with `LOGIN_NODES_STOP` update policy to `LOGIN_NODES_POOL_STOP`.
* Moved login node’s `pool-name` tag creation to Pool construct in `login_nodes_stack` so that each pool resource has correct tag.
* Added pool names to security group descriptions.
* Updated `CHANGELOG`

#### Manual Testing:

* Created cluster containing up to five login node pools with various node counts and instance types.
    * Ensured able to connect with DCV on each pool only when configured
    * Tested slurm job submission for each pool
    * Verified managed security groups for each pool’s nodes and load balancer
* Update cluster
    * Verified new `LOGIN_NODES_POOL_STOP` update policy only applies to pool in which update is requested
    * Verified able to add pools 
    * Verified only able to remove pools when all login nodes are stopped

#### Unit tests

* Updated `test_cluster_operations_controller.py`, `test_update_policy.py`, `test_login_nodes_status.py`, `test_cluster_schema.py` to test multiple pools.
* Removed `pool_validator` unit test

#### Integration tests
* Updated `test_login_node_security_groups` to test security groups of multiple pools.

### References
* Link to impacted open issues.
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/api/controllers/cluster_operations_controller.py [hunk 1]
```diff
@@ -265,18 +265,27 @@ def describe_cluster(cluster_name, region=None):
...
+            pool = LoginNodesPool(status=status)
+            # pool.name = pool_name
+            pool.address = pool_status.get_address()
+            pool.scheme = pool_status.get_scheme()
+            pool.healthy_nodes = pool_status.get_healthy_nodes()
+            pool.unhealthy_nodes = pool_status.get_unhealthy_nodes()
+            login_nodes.append(pool)
+            break
+
+        return login_nodes[0]
```
Comment (hgreebe) [inline cli/src/pcluster/api/controllers/cluster_operations_controller.py:288]: Right now this only returns the first pool right?

File: cli/src/pcluster/config/update_policy.py [hunk 2]
```diff
@@ -491,6 +491,21 @@ def condition_checker_login_nodes_stop_policy(_, patch):
     return not patch.cluster.has_running_login_nodes()
 
 
+def condition_checker_login_nodes_pool_stop_policy(change, patch):
+    """Check if login nodes are running in the pool in which update was requested."""
+    pool_name = get_pool_name_from_change_paths(change)
+    return not patch.cluster.has_running_login_nodes(pool_name=pool_name)
+
+
+def get_pool_name_from_change_paths(change):
```
Comment (hgreebe) [inline cli/src/pcluster/config/update_policy.py:500]: Could you include a comment that shows an example path. Similar to what is done for `get_q_from_config`

File: cli/src/pcluster/config/update_policy.py [hunk 3]
```diff
@@ -679,7 +696,7 @@ def condition_checker_login_nodes_stop_policy(_, patch):
     action_needed=UpdatePolicy.ACTIONS_NEEDED["login_nodes_stop"],
 )
 
-# Update supported only with all login nodes down
+# Update supported only with all login nodes in the cluster down
 UpdatePolicy.LOGIN_NODES_STOP = UpdatePolicy(
```
Comment (hgreebe) [inline cli/src/pcluster/config/update_policy.py:704]: why is this policy needed if we have LOGIN_NODES_POOL_STOP? Is there any update policy for changes to a pool, such as increasing the number of nodes in a pool?

File: cli/src/pcluster/models/cluster.py [hunk 4]
```diff
@@ -755,13 +754,21 @@ def has_running_capacity(self, updated_value: bool = False) -> bool:
...
         return self.__has_running_capacity
 
-    def has_running_login_nodes(self, updated_value: bool = False) -> bool:
-        """Return True if the cluster has running login nodes. Note: the value will be cached."""
+    def has_running_login_nodes(self, updated_value: bool = False, pool_name: str = None) -> bool:
+        """
+        Return True if the cluster has running login nodes, or a specific pool if a pool name is provided.
+
+        Note: the value will be cached.
+        """
```
Comment (hgreebe) [inline cli/src/pcluster/models/cluster.py:762]: Nitpick: Since `get_healthy_nodes` already checks in pool_name in None, you don't need to check it here. You could just call `get_healthy_nodes(pool_name)` regardless if pool_name = None.

File: cli/tests/pcluster/example_configs/slurm.full.yaml [hunk 5]
```diff
@@ -6,8 +6,8 @@ Image:
   CustomAmi: ami-12345678
 LoginNodes:
   Pools:
-  - Name: test
-    InstanceType: t2.micro
+  - Name: pool1
+    InstanceType: c5.large
```
Comment (hgreebe) [inline cli/tests/pcluster/example_configs/slurm.full.yaml]: why the change to `c5.large`

File: cli/tests/pcluster/models/test_login_nodes_status.py [hunk 6]
```diff
@@ -103,23 +132,44 @@ def test_full_login_nodes_status(self, mocker):
...
-            return_value=[self.dummy_load_balancer_2, self.dummy_load_balancers],
+            return_value=[self.dummy_load_balancer_1, self.dummy_load_balancer_2, self.dummy_load_balancer_3],
         )
         mocker.patch("pcluster.aws.elb.ElbClient.describe_tags", return_value=self.dummy_tags_description)
         mocker.patch("pcluster.aws.elb.ElbClient.describe_target_groups", return_value=self.dummy_target_groups)
         mocker.patch("pcluster.aws.elb.ElbClient.describe_target_health", return_value=self.dummy_targets_health)
+
         login_nodes_status = LoginNodesStatus(self.dummy_stack_name)
-        login_nodes_status.retrieve_data(self.dummy_pool_name)
+        login_nodes_status.retrieve_data([self.dummy_pool_name_1, self.dummy_pool_name_2])
```
Comment (hgreebe) [inline cli/tests/pcluster/models/test_login_nodes_status.py]: you should include a check of `login_nodes_status.get_healthy_nodes(pool1)`

File: tests/integration-tests/tests/networking/test_security_groups.py [hunk 7]
```diff
@@ -176,75 +176,122 @@ def test_login_node_security_groups(
...
-            if ip_permissions.get("UserIdGroupPairs"):
-                assert_that(
-                    any(
-                        load_balancer_managed_security_group.get("GroupId") == id_group_pairs.get("GroupId")
-                        for id_group_pairs in ip_permissions.get("UserIdGroupPairs")
-                    )
-                ).is_true()
+        pool_2_load_balancer_managed_sg, pool_2_node_managed_sg = _get_pool_managed_security_groups(
+            "pool2", security_groups
+        )
```
Comment (hgreebe) [inline tests/integration-tests/tests/networking/test_security_groups.py:201]: you could use a for loop that goes through the 2 pools so that you don't have to write each assert twice

File: cli/src/pcluster/api/controllers/cluster_operations_controller.py (hunk 1 above)
Comment (cjmakin) [reply cli/src/pcluster/api/controllers/cluster_operations_controller.py:288]: Yes

File: cli/src/pcluster/config/update_policy.py (hunk 3 above)
Comment (cjmakin) [reply cli/src/pcluster/config/update_policy.py:704]: It's still used [here](https://github.com/aws/aws-parallelcluster/blob/develop/cli/src/pcluster/schemas/cluster_schema.py#L1922) to update the entire login nodes section 

The update policy for changes to a pool is now LOGIN_NODES_POOL_STOP

File: cli/src/pcluster/models/cluster.py (hunk 4 above)
Comment (cjmakin) [reply cli/src/pcluster/models/cluster.py:762]: Nice, thank you!

File: cli/tests/pcluster/models/test_login_nodes_status.py (hunk 6 above)
Comment (cjmakin) [reply cli/tests/pcluster/models/test_login_nodes_status.py]: checks added

File: cli/tests/pcluster/models/test_login_nodes_status.py [hunk 8]
```diff
@@ -103,23 +132,50 @@ def test_full_login_nodes_status(self, mocker):
...
+        pool_2_status = login_nodes_status.get_pool_status_dict().get(self.dummy_pool_name_2)
+
         assert_that(login_nodes_status.get_login_nodes_pool_available()).is_true()
-        assert_that(login_nodes_status.get_status()).is_equal_to(LoginNodesPoolState.ACTIVE)
-        assert_that(login_nodes_status.get_address()).is_equal_to(self.dummy_dns_name)
-        assert_that(login_nodes_status.get_scheme()).is_equal_to(self.dummy_scheme)
-        assert_that(login_nodes_status.get_healthy_nodes()).is_equal_to(2)
-        assert_that(login_nodes_status.get_unhealthy_nodes()).is_equal_to(1)
+
+        assert_that(pool_1_status.get_status()).is_equal_to(LoginNodesPoolState.ACTIVE)
```
Comment (hgreebe) [inline cli/tests/pcluster/models/test_login_nodes_status.py]: Nitpick: I would also probably recommend doing these checks in a for loop.

**Review by hgreebe — DISMISSED** [review]

**Review by hgreebe — APPROVED** [review]
